tarsier/tag_utils.ts

```diff
@@ -458,67 +461,124 @@ function insertTags(
           (child as HTMLElement).tagName.toLowerCase(),
         )
       ) {
-        return getElementToInsertInto(child as HTMLElement, idSpan);
+        return getElementToInsertInto(child as HTMLElement);
       }
     }
 
     trimTextNodeStart(element);
-
-    // Base case
     return element;
   }
 
-  const idToTagMetadata: { [key: number]: TagMetadata } = {};
-  let idNum = 0;
-
-  for (let el of elementsToTag) {
-    idToTagMetadata[idNum] = {
-      xpath: getElementXPath(el),
-    };
+  function getOpeningTag(el: HTMLElement): string {
+    return el.outerHTML.split(">")[0] + ">";
+  }
 
-    if (isInteractable(el)) {
-      const idSpan = create_tagged_span(idNum, el);
-      if (isTextInsertable(el) && el.parentElement) {
-        el.parentElement.insertBefore(idSpan, el);
-      } else {
-        const insertionElement = getElementToInsertInto(el, idSpan);
-        insertionElement.prepend(idSpan);
+  function createAndInsertTag(
+    el: HTMLElement,
+    xpath: string,
+    textNodeIndex: number | null,
+    isAbsolutelyPositioned: boolean,
+    referenceNode: ChildNode | null = null,
+    originalTextContent: string | null = null,
+  ) {
+    const symbol = getTagSymbol(el);
+    const idSpan = create_tagged_span(idNum, symbol, el);
+
+    tagDataList.push({
+      xpath,
+      element: el,
+      tagElement: idSpan,
+      textNodeIndex,
+      originalTextContent,
+    });
```

Comment:
a bit sus to use a closure variable inside this inlined function before the variables even been created. These functions should all be pure functions. We should make it return the tagDataList entry if anything